proj1a/LinkedListDeque.java

```diff
@@ -0,0 +1,170 @@
+public class LinkedListDeque<T> implements Deque<T> {
+    private class LLDNode {
+        LLDNode prev;
+        T item;
+        LLDNode next;
+
+        public LLDNode(LLDNode front, T n, LLDNode back) {
+            prev = front;
+            item = n;
+            next = back;
+        }
+    }
+
+    private LLDNode sentinel;
+    private int size;
+
+    /**
+     * create empty linked list deque
+     */
+    public LinkedListDeque() {
+        sentinel = new LLDNode (null, null, null);
+        sentinel.prev = sentinel;
+        sentinel.next = sentinel;
+        size = 0;
+    }
+
+    /**
+     * add item to front of the deque
+     * @param item: item to add
+     */
+    @Override
+    public synchronized void addFirst(T item) {
+        LLDNode temp = new LLDNode(sentinel, item, sentinel.next);
+        sentinel.next.prev = temp;
+        sentinel.next = temp;
+        size = size + 1;
+    }
+
+    /**
+     * add item to back of the deque
+     * @param item: item to add
+     */
+    @Override
+    public synchronized void addLast(T item) {
+        LLDNode temp = new LLDNode(sentinel.prev, item, sentinel);
+        sentinel.prev.next = temp;
+        sentinel.prev = temp;
+        size = size + 1;
+    }
+
+    /**
+     * return if deque is empty
+     */
+    @Override
+    public boolean isEmpty() {
+        return size == 0;
+    }
+
+    /**
+     * number of items in the deque
+     */
+    @Override
+    public int size() {
+        return size;
+    }
+
+    /**
+     * print items in the deque
+     */
+    @Override
+    public void printDeque() {
+        LLDNode p = sentinel.next;
+        while (p != sentinel) {
+            System.out.print(p.item + " ");
+            p = p.next;
+        }
+        System.out.println();
+    }
+
+    /**
+     * remove and return item at front of deque
+     */
+    @Override
+    public synchronized T removeFirst() {
+        //if no first item
+        if (isEmpty()) {
+            return null;
+        }
+
+        T result = sentinel.next.item;
+        LLDNode target = sentinel.next.next;
+        sentinel.next = target;
+        size = size - 1;
```

Comment:
What about the target.prev that still points to the removed node? Before fixing it, could you add a unit test that would find this bug?

Reply:
Bug fixed, unit test added in ```LinkedListTest.java```